lib/trip.rb

```diff
@@ -1,71 +1,99 @@
 require 'csv'
+require 'time'
 
 require_relative 'csv_record'
 
 module RideShare
   class Trip < CsvRecord
-    attr_reader :id, :passenger, :passenger_id, :start_time, :end_time, :cost, :rating
+    attr_reader :id, :passenger, :passenger_id, :driver, :driver_id, :start_time, :end_time, :cost, :rating
 
     def initialize(
           id:,
+          driver: nil,
+          driver_id: nil,
           passenger: nil,
           passenger_id: nil,
           start_time:,
           end_time:,
           cost: nil,
           rating:
         )
-      super(id)
+      super(id) #super class csv record inherits from csv initialize method self.class.validate_id(id)
+
+      #When a Trip is constructed, either driver_id or driver must be provided.
+      if driver
+        @driver = driver
+        @driver_id = driver.id
+      elsif driver_id
+        @driver_id = driver_id
+      else
+        raise ArgumentError, 'Driver or driver_id is required'
+      end
 
       if passenger
         @passenger = passenger
         @passenger_id = passenger.id
-
       elsif passenger_id
         @passenger_id = passenger_id
-
       else
         raise ArgumentError, 'Passenger or passenger_id is required'
       end
 
       @start_time = start_time
       @end_time = end_time
+
+
+      raise ArgumentError, "End time cannot be before start time." unless @end_time.nil? || @end_time > @start_time
+
       @cost = cost
       @rating = rating
 
-      if @rating > 5 || @rating < 1
-        raise ArgumentError.new("Invalid rating #{@rating}")
+      unless rating.nil?
+        if @rating > 5 || @rating < 1
+          raise ArgumentError.new("Invalid rating #{@rating}")
+        end
```

Comment:
On potential refactor to this code would be to make a method for the validation of each parameter. This primarily serves to reduce clutter in the constructors and make the code more readable.